tools/tpm2_key_import.c

```diff
@@ -0,0 +1,578 @@
+#include <openssl/sha.h>
+#include <openssl/hmac.h>
+#include <openssl/aes.h>
+#include <openssl/rsa.h>
+
+#include <getopt.h>
+#include <limits.h>
+#include <sapi/tpm20.h>
+#include <sapi//marshal.h>
+
+#include "log.h"
+#include "files.h"
+#include "tpm_kdfa.h"
+#include "main.h"
+#include "options.h"
+#include "tpm2_util.h"
+
+#define SYM_KEY_SIZE 16
+#define max_buffer_size  1024
+
+typedef struct tpm_import_ctx tpm_import_ctx;
+struct tpm_import_ctx {
+    char *input_key_file;
+    char *import_key_public_file;
+    char *import_key_private_file;
+    char *parent_key_public_file;
+    uint8_t *input_key_buffer;
+    //Parent public key for seed encryption
+    TPM2B_PUBLIC_KEY_RSA parent_public_key;
+    TPM_HANDLE parent_key_handle;
+    //External key public
+    TPM2B_PUBLIC import_key_public;
+    //External key name
+    TPM2B_NAME import_key_public_name;
+    //External key sensitive
+    TPM2B_SENSITIVE import_key_sensitive;
+    //External key private
+    TPM2B_PRIVATE import_key_private;
+    //Protection Seed and keys
+    uint8_t protection_seed_data[SHA256_DIGEST_SIZE]; //max tpm digest
+    uint8_t encrypted_protection_seed_data[MAX_RSA_KEY_BYTES];
+    uint8_t protection_hmac_key[SHA256_DIGEST_SIZE];
+    uint8_t protection_enc_key[SYM_KEY_SIZE];
+    uint8_t import_key_public_unique_data[SHA256_DIGEST_SIZE];
+    uint8_t outer_integrity_hmac[SHA256_DIGEST_SIZE];
+    TPM2B_DATA enc_sensitive_key;
+    TPM2B_MAX_BUFFER encrypted_inner_integrity;
+    TPM2B_MAX_BUFFER encrypted_duplicate_sensitive;
+    //SAPI and TCTI context
+    TSS2_SYS_CONTEXT *sapi_context;
+};
+
+int ssl_RSA_set0_key(RSA *r, BIGNUM *n, BIGNUM *e, BIGNUM *d) {
+
+    if ((r->n == NULL && n == NULL) || (r->e == NULL && e == NULL))
+        return 0;
+    if (n != NULL) {
+        BN_free(r->n);
+        r->n = n;
+    }
+    if (e != NULL) {
+        BN_free(r->e);
+        r->e = e;
+    }
+    if (d != NULL) {
+        BN_free(r->d);
+        r->d = d;
+    }
+    return 1;
+}
+
+bool encrypt_seed_with_tpm2_rsa_public_key(tpm_import_ctx *ctx) {
+    //Public Modulus
+    FILE *fp = fopen(ctx->parent_key_public_file, "rb");
+    if (fp == NULL) {
+        LOG_ERR("Failed accessing parent key public file.");
+        return false;
+    }
+    if (fseek(fp, 102, SEEK_SET) != 0) {
+        LOG_ERR("Expected parent key public data file size failure");
+        return false;
+    }
+    unsigned char pub_modulus[MAX_RSA_KEY_BYTES] = { 0 };
+    int ret = fread(pub_modulus, 1, MAX_RSA_KEY_BYTES, fp);
+    if (ret != MAX_RSA_KEY_BYTES) {
+        LOG_ERR("Failed reading public modulus from parent key public file");
+        return false;
+    }
+    fclose(fp);
+    RSA *rsa = NULL;
+    unsigned char encoded[MAX_RSA_KEY_BYTES];
+    unsigned char label[10] = { 'D', 'U', 'P', 'L', 'I', 'C', 'A', 'T', 'E', 0 };
+    int return_code = RSA_padding_add_PKCS1_OAEP_mgf1(encoded,
+            MAX_RSA_KEY_BYTES, ctx->protection_seed_data, 32, label, 10,
```

Comment:
why 32? size of sha256 digest in bytes?
Can't 10 be replaced with sizeof(label)?